File: Python/test/TrieTree.py
class TrieNode:
    def __init__(self):
        self.pas, self.end = 0, 0
        self.nexts = {}


class TrieTree:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        cur = self.root
        cur.pas += 1
        for c in word:
            if c not in cur.nexts:
                cur.nexts[c] = TrieNode()
            cur = cur.nexts[c]
            cur.pas += 1
        cur.end += 1  # 若是word的最后一个字符，end+=1

    def search(self, word):
        """
        查询 word 加入的次数
        """
        cur = self.root
        for c in word:
            if c not in cur.nexts:
                return 0
            cur = cur.nexts[c]
        return cur.end

    def delete(self, word):
        if self.search(word) > 0:  # 删除之前先看看word在不在树中
            # 删除方式是，沿途pas-=1
            cur = self.root
            cur.pas -= 1
            for c in word:
                cur.nexts[c].pas -= 1
                if cur.nexts[c].pas == 0:  # 检查下方节点是不是空了
                    cur.nexts[c] = None  # 直接修改父节点的指针为空
                    return
                cur = cur.nexts[c]
            cur.end -= 1  # 到最后一个字符还要end-=1
        print("Word to be deleted does not exist.")

    def prefix_num(self, pre_string):
        """
        所有加入的字符串，有多少是以pre_string为前缀
        """
        cur = self.root
        for c in pre_string:
            if c not in cur.nexts:
                return 0
            cur = cur.nexts[c]
        return cur.pas


if __name__ == '__main__':
    tree = TrieTree()
    word_list = ['abc', 'abcde', 'badd', 'ab', 'a', 'abbd', 'abc']
    for w in word_list:
        tree.insert(w)
    print(tree.search('abc'))
    print(tree.prefix_num('a'))
    print(tree.prefix_num('ab'))
    print(tree.prefix_num('abc'))
    tree.delete('ab')
    tree.delete('bb')
    print(tree.search('ab'))

File: Java/src/test/TrieTree.java
package test;

import java.util.HashMap;
import java.util.Map;

/**
 * 前缀树定义
 */
public class TrieTree {
    TrieNode root;

    public TrieTree() {
        root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode cur = root;
        cur.pas++;
        for (char c : word.toCharArray()) {
            // 如果cur.nexts没有里存着当前字符，则新建节点
            if (!cur.nexts.containsKey(c)) {
                cur.nexts.put(c, new TrieNode());
            }
            cur = cur.nexts.get(c);
            cur.pas++;
        }
        cur.end++;
    }

    public int search(String word) {
        TrieNode cur = root;
        for (char c : word.toCharArray()) {
            cur = cur.nexts.get(c);
            if (cur == null) {
                return 0;
            }
        }
        return cur.end;
    }

    public void delete(String word) {
        if (search(word) > 0) { // 有这个单词才做删除
            TrieNode cur = root;
            cur.pas--;
            for (char c : word.toCharArray()) {
                cur = cur.nexts.get(c);
                cur.pas--;
                if (cur.pas == 0) {
                    // 已经没有单词经过这个节点了，
                    // 那么直接删除该节点（以及以下的内容），完成删除
                    cur.nexts.remove(c);
                    return;
                }
            }
            cur.end--;
        }
    }

    /**
     * 求以preString为前缀的字符串数量
     *
     * @param preString 前缀字符串
     * @return 数量
     */
    public int prefixNum(String preString) {
        TrieNode cur = root;
        for (char c : preString.toCharArray()) {
            cur = cur.nexts.get(c);
            if (cur == null) {
                return 0;
            }
        }
        return cur.pas;
    }


    public static void main(String[] args) {
        String[] strings = new String[]{"abc", "abd", "abcd", "aaa", "cbd", "aaa"};
        TrieTree trieTree = new TrieTree();
        for (String s : strings) {
            trieTree.insert(s);
        }
        System.out.println(trieTree.search("aew"));
        //System.out.println(trieTree.prefixNum("ab"));
    }

}


/**
 * 节点定义
 */
class TrieNode {
    int pas;  // 通过的次数
    int end;  // 以该节点为结尾的次数
    Map<Character, TrieNode> nexts;

    public TrieNode() {
        pas = 0;
        end = 0;
        nexts = new HashMap<>();
    }
}
